From 59bb4642e2a8e99c31794333fe5b37386ac5faf7 Mon Sep 17 00:00:00 2001
From: Doug Semler <dougsemler@gmail.com>
Date: Mon, 24 Jul 2017 14:01:02 +0300
Subject: [PATCH 2/5] Fix PE-COFF archiving long name tables

Fixes the PE-COFF archive long name tables by creating a new set of
functions in the bfd library and using those functions when building
PE-COFF archives.  By default, the PE-COFF archive functions are defined
to be the normal COFF archive functions.

	* libbfd-in.h: Declare _bfd_archive_pecoff* functions.
	* libbfd.h: Regenerate.
	* archive.c (_bfd_archive_pecoff_construct_extended_name_table):
	  New.
	* coff-i386.c: Use PE-COFF archive functions when COFF_WITH_PE.
	* coff-ia64.c: Likewise.
	* coff-ppc.c: Likewise.
	* coff-sh.c: Likewise.
	* coff-x86_64.c: Likewise.
	* coffcode.h: Likewise.
	* pe-mips.c: Likewise.
---
 bfd/archive.c     |  28 ++++++++++
 bfd/coff-i386.c   |   4 ++
 bfd/coff-ia64.c   |   4 ++
 bfd/coff-ppc.c    |   8 +++
 bfd/coff-sh.c     |   8 +++
 bfd/coff-x86_64.c |   4 ++
 bfd/coffcode.h    | 159 ++++++++++++++++++++++++++++++++++++++++++++++++++++++
 bfd/libbfd-in.h   |  23 ++++++++
 bfd/libbfd.h      |  23 ++++++++
 bfd/pe-mips.c     |   4 ++
 10 files changed, 265 insertions(+)

diff --git a/bfd/archive.c b/bfd/archive.c
index f209bab..51386b9 100644
--- a/bfd/archive.c
+++ b/bfd/archive.c
@@ -1580,6 +1580,34 @@ _bfd_archive_coff_construct_extended_name_table (bfd *abfd,
   return _bfd_construct_extended_name_table (abfd, TRUE, tabloc, tablen);
 }
 
+/* PE-COFF long filenames are null terminated, not "/\012", which is
+   what the coff archiver function does.  So we call general function
+   without asking for terminating slashes.  Then, we replace
+   all of the '\012' characters with '\0'.
+   Information obtained from the PE-COFF Specification version 8.1,
+   section 7.5 on the Longnames member. */
+
+bfd_boolean
+_bfd_archive_pecoff_construct_extended_name_table (bfd *abfd,
+						   char **tabloc,
+						   bfd_size_type *tablen,
+						   const char **name)
+{
+  bfd_size_type i;
+  char *strptr;
+  *name = "//";
+  if (! _bfd_construct_extended_name_table (abfd, FALSE, tabloc, tablen))
+    return FALSE;
+
+  strptr = *tabloc;
+  for (i = 0; i < *tablen; ++ i)
+    {
+      if (strptr[i] == ARFMAG[1])
+        strptr[i] = '\0';
+    }
+  return TRUE;
+}
+
 /* Follows archive_head and produces an extended name table if
    necessary.  Returns (in tabloc) a pointer to an extended name
    table, and in tablen the length of the table.  If it makes an entry
diff --git a/bfd/coff-i386.c b/bfd/coff-i386.c
index b6ef597..878b209 100644
--- a/bfd/coff-i386.c
+++ b/bfd/coff-i386.c
@@ -671,7 +671,11 @@ const bfd_target
      BFD_JUMP_TABLE_GENERIC (coff),
      BFD_JUMP_TABLE_COPY (coff),
      BFD_JUMP_TABLE_CORE (_bfd_nocore),
+#ifndef COFF_WITH_PE
      BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_coff),
+#else
+     BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_pecoff),
+#endif
      BFD_JUMP_TABLE_SYMBOLS (coff),
      BFD_JUMP_TABLE_RELOCS (coff),
      BFD_JUMP_TABLE_WRITE (coff),
diff --git a/bfd/coff-ia64.c b/bfd/coff-ia64.c
index fe27350..efce78b 100644
--- a/bfd/coff-ia64.c
+++ b/bfd/coff-ia64.c
@@ -189,7 +189,11 @@ const bfd_target
      BFD_JUMP_TABLE_GENERIC (coff),
      BFD_JUMP_TABLE_COPY (coff),
      BFD_JUMP_TABLE_CORE (_bfd_nocore),
+#ifndef COFF_WITH_PE
      BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_coff),
+#else
+     BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_pecoff),
+#endif
      BFD_JUMP_TABLE_SYMBOLS (coff),
      BFD_JUMP_TABLE_RELOCS (coff),
      BFD_JUMP_TABLE_WRITE (coff),
diff --git a/bfd/coff-ppc.c b/bfd/coff-ppc.c
index 9dca30c..f2ef9c3 100644
--- a/bfd/coff-ppc.c
+++ b/bfd/coff-ppc.c
@@ -2517,7 +2517,11 @@ const bfd_target TARGET_LITTLE_SYM =
   BFD_JUMP_TABLE_GENERIC (coff),
   BFD_JUMP_TABLE_COPY (coff),
   BFD_JUMP_TABLE_CORE (_bfd_nocore),
+#ifndef COFF_WITH_PE
   BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_coff),
+#else
+  BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_pecoff),
+#endif
   BFD_JUMP_TABLE_SYMBOLS (coff),
   BFD_JUMP_TABLE_RELOCS (coff),
   BFD_JUMP_TABLE_WRITE (coff),
@@ -2578,7 +2582,11 @@ const bfd_target TARGET_BIG_SYM =
   BFD_JUMP_TABLE_GENERIC (coff),
   BFD_JUMP_TABLE_COPY (coff),
   BFD_JUMP_TABLE_CORE (_bfd_nocore),
+#ifndef COFF_WITH_PE
   BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_coff),
+#else
+  BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_pecoff),
+#endif
   BFD_JUMP_TABLE_SYMBOLS (coff),
   BFD_JUMP_TABLE_RELOCS (coff),
   BFD_JUMP_TABLE_WRITE (coff),
diff --git a/bfd/coff-sh.c b/bfd/coff-sh.c
index 26d296b..6a62f83 100644
--- a/bfd/coff-sh.c
+++ b/bfd/coff-sh.c
@@ -3156,7 +3156,11 @@ const bfd_target sh_coff_small_vec =
   BFD_JUMP_TABLE_GENERIC (coff_small),
   BFD_JUMP_TABLE_COPY (coff),
   BFD_JUMP_TABLE_CORE (_bfd_nocore),
+#ifndef COFF_WITH_PE
   BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_coff),
+#else
+  BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_pecoff),
+#endif
   BFD_JUMP_TABLE_SYMBOLS (coff),
   BFD_JUMP_TABLE_RELOCS (coff),
   BFD_JUMP_TABLE_WRITE (coff),
@@ -3201,7 +3205,11 @@ const bfd_target sh_coff_small_le_vec =
   BFD_JUMP_TABLE_GENERIC (coff_small),
   BFD_JUMP_TABLE_COPY (coff),
   BFD_JUMP_TABLE_CORE (_bfd_nocore),
+#ifndef COFF_WITH_PE
   BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_coff),
+#else
+  BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_pecoff),
+#endif
   BFD_JUMP_TABLE_SYMBOLS (coff),
   BFD_JUMP_TABLE_RELOCS (coff),
   BFD_JUMP_TABLE_WRITE (coff),
diff --git a/bfd/coff-x86_64.c b/bfd/coff-x86_64.c
index de22822..3b8c43a 100644
--- a/bfd/coff-x86_64.c
+++ b/bfd/coff-x86_64.c
@@ -804,7 +804,11 @@ const bfd_target
   BFD_JUMP_TABLE_GENERIC (coff),
   BFD_JUMP_TABLE_COPY (coff),
   BFD_JUMP_TABLE_CORE (_bfd_nocore),
+#ifndef COFF_WITH_PE
   BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_coff),
+#else
+  BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_pecoff),
+#endif
   BFD_JUMP_TABLE_SYMBOLS (coff),
   BFD_JUMP_TABLE_RELOCS (coff),
   BFD_JUMP_TABLE_WRITE (coff),
diff --git a/bfd/coffcode.h b/bfd/coffcode.h
index 4aa9741..4f71ff6 100644
--- a/bfd/coffcode.h
+++ b/bfd/coffcode.h
@@ -6064,6 +6064,10 @@ static bfd_coff_backend_data bigobj_swap_table =
 #define coff_bfd_define_start_stop	    bfd_generic_define_start_stop
 #endif
 
+/* Define the PE-COFF generators with the correct tables. */
+
+#ifndef COFF_WITH_PE
+
 #define CREATE_BIG_COFF_TARGET_VEC(VAR, NAME, EXTRA_O_FLAGS, EXTRA_S_FLAGS, UNDER, ALTERNATIVE, SWAP_TABLE)	\
 const bfd_target VAR =							\
 {									\
@@ -6214,3 +6218,158 @@ const bfd_target VAR =							\
 									\
   SWAP_TABLE								\
 };
+
+#else /* COFF_WITH_PE */
+
+#define CREATE_BIG_COFF_TARGET_VEC(VAR, NAME, EXTRA_O_FLAGS, EXTRA_S_FLAGS, UNDER, ALTERNATIVE, SWAP_TABLE)	\
+const bfd_target VAR =							\
+{									\
+  NAME ,								\
+  bfd_target_coff_flavour,						\
+  BFD_ENDIAN_BIG,		/* Data byte order is big.  */		\
+  BFD_ENDIAN_BIG,		/* Header byte order is big.  */	\
+  /* object flags */							\
+  (HAS_RELOC | EXEC_P | HAS_LINENO | HAS_DEBUG |			\
+   HAS_SYMS | HAS_LOCALS | WP_TEXT | EXTRA_O_FLAGS),			\
+  /* section flags */							\
+  (SEC_HAS_CONTENTS | SEC_ALLOC | SEC_LOAD | SEC_RELOC | EXTRA_S_FLAGS),\
+  UNDER,			/* Leading symbol underscore.  */	\
+  '/',				/* AR_pad_char.  */			\
+  15,				/* AR_max_namelen.  */			\
+  0,				/* match priority.  */			\
+									\
+  /* Data conversion functions.  */					\
+  bfd_getb64, bfd_getb_signed_64, bfd_putb64,				\
+  bfd_getb32, bfd_getb_signed_32, bfd_putb32,				\
+  bfd_getb16, bfd_getb_signed_16, bfd_putb16,				\
+									\
+  /* Header conversion functions.  */					\
+  bfd_getb64, bfd_getb_signed_64, bfd_putb64,				\
+  bfd_getb32, bfd_getb_signed_32, bfd_putb32,				\
+  bfd_getb16, bfd_getb_signed_16, bfd_putb16,				\
+									\
+	/* bfd_check_format.  */					\
+  { _bfd_dummy_target, coff_object_p, bfd_generic_archive_p,		\
+    _bfd_dummy_target },						\
+	/* bfd_set_format.  */						\
+  { bfd_false, coff_mkobject, _bfd_generic_mkarchive, bfd_false },	\
+	/* bfd_write_contents.  */					\
+  { bfd_false, coff_write_object_contents, _bfd_write_archive_contents,	\
+    bfd_false },							\
+									\
+  BFD_JUMP_TABLE_GENERIC (coff),					\
+  BFD_JUMP_TABLE_COPY (coff),						\
+  BFD_JUMP_TABLE_CORE (_bfd_nocore),					\
+  BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_pecoff),				\
+  BFD_JUMP_TABLE_SYMBOLS (coff),					\
+  BFD_JUMP_TABLE_RELOCS (coff),						\
+  BFD_JUMP_TABLE_WRITE (coff),						\
+  BFD_JUMP_TABLE_LINK (coff),						\
+  BFD_JUMP_TABLE_DYNAMIC (_bfd_nodynamic),				\
+									\
+  ALTERNATIVE,								\
+									\
+  SWAP_TABLE								\
+};
+
+#define CREATE_BIGHDR_COFF_TARGET_VEC(VAR, NAME, EXTRA_O_FLAGS, EXTRA_S_FLAGS, UNDER, ALTERNATIVE, SWAP_TABLE)	\
+const bfd_target VAR =							\
+{									\
+  NAME ,								\
+  bfd_target_coff_flavour,						\
+  BFD_ENDIAN_LITTLE,		/* Data byte order is little.  */	\
+  BFD_ENDIAN_BIG,		/* Header byte order is big.  */	\
+  /* object flags */							\
+  (HAS_RELOC | EXEC_P | HAS_LINENO | HAS_DEBUG |			\
+   HAS_SYMS | HAS_LOCALS | WP_TEXT | EXTRA_O_FLAGS),			\
+  /* section flags */							\
+  (SEC_HAS_CONTENTS | SEC_ALLOC | SEC_LOAD | SEC_RELOC | EXTRA_S_FLAGS),\
+  UNDER,			/* Leading symbol underscore.  */	\
+  '/',				/* AR_pad_char.  */			\
+  15,				/* AR_max_namelen.  */			\
+  0,				/* match priority.  */			\
+									\
+  /* Data conversion functions.  */					\
+  bfd_getb64, bfd_getb_signed_64, bfd_putb64,				\
+  bfd_getb32, bfd_getb_signed_32, bfd_putb32,				\
+  bfd_getb16, bfd_getb_signed_16, bfd_putb16,				\
+									\
+  /* Header conversion functions.  */					\
+  bfd_getb64, bfd_getb_signed_64, bfd_putb64,				\
+  bfd_getb32, bfd_getb_signed_32, bfd_putb32,				\
+  bfd_getb16, bfd_getb_signed_16, bfd_putb16,				\
+									\
+	/* bfd_check_format.  */					\
+  { _bfd_dummy_target, coff_object_p, bfd_generic_archive_p,		\
+    _bfd_dummy_target },						\
+	/* bfd_set_format.  */						\
+  { bfd_false, coff_mkobject, _bfd_generic_mkarchive, bfd_false },	\
+	/* bfd_write_contents.  */					\
+  { bfd_false, coff_write_object_contents, _bfd_write_archive_contents,	\
+    bfd_false },							\
+									\
+  BFD_JUMP_TABLE_GENERIC (coff),					\
+  BFD_JUMP_TABLE_COPY (coff),						\
+  BFD_JUMP_TABLE_CORE (_bfd_nocore),					\
+  BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_pecoff),				\
+  BFD_JUMP_TABLE_SYMBOLS (coff),					\
+  BFD_JUMP_TABLE_RELOCS (coff),						\
+  BFD_JUMP_TABLE_WRITE (coff),						\
+  BFD_JUMP_TABLE_LINK (coff),						\
+  BFD_JUMP_TABLE_DYNAMIC (_bfd_nodynamic),				\
+									\
+  ALTERNATIVE,								\
+									\
+  SWAP_TABLE								\
+};
+
+#define CREATE_LITTLE_COFF_TARGET_VEC(VAR, NAME, EXTRA_O_FLAGS, EXTRA_S_FLAGS, UNDER, ALTERNATIVE, SWAP_TABLE)	\
+const bfd_target VAR =							\
+{									\
+  NAME ,								\
+  bfd_target_coff_flavour,						\
+  BFD_ENDIAN_LITTLE,		/* Data byte order is little.  */	\
+  BFD_ENDIAN_LITTLE,		/* Header byte order is little.  */	\
+	/* object flags */						\
+  (HAS_RELOC | EXEC_P | HAS_LINENO | HAS_DEBUG |			\
+   HAS_SYMS | HAS_LOCALS | WP_TEXT | EXTRA_O_FLAGS),			\
+	/* section flags */						\
+  (SEC_HAS_CONTENTS | SEC_ALLOC | SEC_LOAD | SEC_RELOC | EXTRA_S_FLAGS),\
+  UNDER,			/* Leading symbol underscore.  */	\
+  '/',				/* AR_pad_char.  */			\
+  15,				/* AR_max_namelen.  */			\
+  0,				/* match priority.  */			\
+									\
+  /* Data conversion functions.  */					\
+  bfd_getl64, bfd_getl_signed_64, bfd_putl64,				\
+  bfd_getl32, bfd_getl_signed_32, bfd_putl32,				\
+  bfd_getl16, bfd_getl_signed_16, bfd_putl16,				\
+  /* Header conversion functions.  */					\
+  bfd_getl64, bfd_getl_signed_64, bfd_putl64,				\
+  bfd_getl32, bfd_getl_signed_32, bfd_putl32,				\
+  bfd_getl16, bfd_getl_signed_16, bfd_putl16,				\
+	/* bfd_check_format.  */					\
+  { _bfd_dummy_target, coff_object_p, bfd_generic_archive_p,		\
+    _bfd_dummy_target },						\
+       /* bfd_set_format.  */						\
+  { bfd_false, coff_mkobject, _bfd_generic_mkarchive, bfd_false },	\
+	/* bfd_write_contents.  */					\
+  { bfd_false, coff_write_object_contents, _bfd_write_archive_contents,	\
+    bfd_false },							\
+									\
+  BFD_JUMP_TABLE_GENERIC (coff),					\
+  BFD_JUMP_TABLE_COPY (coff),						\
+  BFD_JUMP_TABLE_CORE (_bfd_nocore),					\
+  BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_pecoff),				\
+  BFD_JUMP_TABLE_SYMBOLS (coff),					\
+  BFD_JUMP_TABLE_RELOCS (coff),						\
+  BFD_JUMP_TABLE_WRITE (coff),						\
+  BFD_JUMP_TABLE_LINK (coff),						\
+  BFD_JUMP_TABLE_DYNAMIC (_bfd_nodynamic),				\
+									\
+  ALTERNATIVE,								\
+									\
+  SWAP_TABLE								\
+};
+
+#endif /* COFF_WITH_PE */
diff --git a/bfd/libbfd-in.h b/bfd/libbfd-in.h
index 7adee99..cbdef8c 100644
--- a/bfd/libbfd-in.h
+++ b/bfd/libbfd-in.h
@@ -351,6 +351,29 @@ extern bfd_boolean _bfd_archive_coff_construct_extended_name_table
   bfd_generic_stat_arch_elt
 #define _bfd_archive_coff_update_armap_timestamp bfd_true
 
+/* Routines to use for BFD_JUMP_TABLE_ARCHIVE to get PE-COFF style
+   archives.  Use BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_pecoff).
+   Unless overridden, they default to the COFF functions.  */
+
+#define _bfd_archive_pecoff_slurp_armap _bfd_archive_coff_slurp_armap
+#define _bfd_archive_pecoff_slurp_extended_name_table \
+  _bfd_archive_coff_slurp_extended_name_table
+extern bfd_boolean _bfd_archive_pecoff_construct_extended_name_table
+  (bfd *, char **, bfd_size_type *, const char **);
+#define _bfd_archive_pecoff_truncate_arname \
+  _bfd_archive_coff_truncate_arname
+#define _bfd_archive_pecoff_write_armap _bfd_archive_coff_write_armap
+#define _bfd_archive_pecoff_read_ar_hdr _bfd_archive_coff_read_ar_hdr
+#define _bfd_archive_pecoff_write_ar_hdr _bfd_archive_coff_write_ar_hdr
+#define _bfd_archive_pecoff_openr_next_archived_file \
+  _bfd_archive_coff_openr_next_archived_file
+#define _bfd_archive_pecoff_get_elt_at_index \
+  _bfd_archive_coff_get_elt_at_index
+#define _bfd_archive_pecoff_generic_stat_arch_elt \
+  _bfd_archive_coff_generic_stat_arch_elt
+#define _bfd_archive_pecoff_update_armap_timestamp \
+  _bfd_archive_coff_update_armap_timestamp
+
 /* Routines to use for BFD_JUMP_TABLE_ARCHIVE to get BSD4.4 style
    archives.  Use BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_bsd44).  */
 
diff --git a/bfd/libbfd.h b/bfd/libbfd.h
index ae9bf76..334da52 100644
--- a/bfd/libbfd.h
+++ b/bfd/libbfd.h
@@ -356,6 +356,29 @@ extern bfd_boolean _bfd_archive_coff_construct_extended_name_table
   bfd_generic_stat_arch_elt
 #define _bfd_archive_coff_update_armap_timestamp bfd_true
 
+/* Routines to use for BFD_JUMP_TABLE_ARCHIVE to get PE-COFF style
+   archives.  Use BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_pecoff).
+   Unless overridden, they default to the COFF functions.  */
+
+#define _bfd_archive_pecoff_slurp_armap _bfd_archive_coff_slurp_armap
+#define _bfd_archive_pecoff_slurp_extended_name_table \
+  _bfd_archive_coff_slurp_extended_name_table
+extern bfd_boolean _bfd_archive_pecoff_construct_extended_name_table
+  (bfd *, char **, bfd_size_type *, const char **);
+#define _bfd_archive_pecoff_truncate_arname \
+  _bfd_archive_coff_truncate_arname
+#define _bfd_archive_pecoff_write_armap _bfd_archive_coff_write_armap
+#define _bfd_archive_pecoff_read_ar_hdr _bfd_archive_coff_read_ar_hdr
+#define _bfd_archive_pecoff_write_ar_hdr _bfd_archive_coff_write_ar_hdr
+#define _bfd_archive_pecoff_openr_next_archived_file \
+  _bfd_archive_coff_openr_next_archived_file
+#define _bfd_archive_pecoff_get_elt_at_index \
+  _bfd_archive_coff_get_elt_at_index
+#define _bfd_archive_pecoff_generic_stat_arch_elt \
+  _bfd_archive_coff_generic_stat_arch_elt
+#define _bfd_archive_pecoff_update_armap_timestamp \
+  _bfd_archive_coff_update_armap_timestamp
+
 /* Routines to use for BFD_JUMP_TABLE_ARCHIVE to get BSD4.4 style
    archives.  Use BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_bsd44).  */
 
diff --git a/bfd/pe-mips.c b/bfd/pe-mips.c
index d746a44..f083433 100644
--- a/bfd/pe-mips.c
+++ b/bfd/pe-mips.c
@@ -911,7 +911,11 @@ const bfd_target
   BFD_JUMP_TABLE_GENERIC (coff),
   BFD_JUMP_TABLE_COPY (coff),
   BFD_JUMP_TABLE_CORE (_bfd_nocore),
+#ifndef COFF_WITH_PE
   BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_coff),
+#else
+  BFD_JUMP_TABLE_ARCHIVE (_bfd_archive_pecoff),
+#endif
   BFD_JUMP_TABLE_SYMBOLS (coff),
   BFD_JUMP_TABLE_RELOCS (coff),
   BFD_JUMP_TABLE_WRITE (coff),
-- 
1.8.5.5

